src/iceberg/type.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+/// \file iceberg/type.h
+/// Data types for Iceberg.
+
+#include <memory>
+#include <span>
+#include <string>
+#include <vector>
+
+#include "iceberg/iceberg_export.h"
+#include "iceberg/type_fwd.h"
+
+namespace iceberg {
+
+/// \brief A type combined with a name.
+class ICEBERG_EXPORT Field {
+ public:
+  /// \brief Construct a field.
+  /// \param[in] field_id The field ID.
+  /// \param[in] name The field name.
+  /// \param[in] type The field type.
+  /// \param[in] optional Whether values of this field are required or nullable.
+  Field(int32_t field_id, std::string name, std::shared_ptr<Type> type, bool optional);
```

Comment:
> std::shared_ptr<Type> type
Does this assume that for each type there is going to be an instance and the Fields will share these instances? Would it be too much overhead to keep a copy of a Type for each Field?

Reply:
Well, I think it's been too long for me since I found myself writing Java code and not C++ on a daily basis :) After taking another look, sharep_ptr<Type> seems fine here. Sorry :)